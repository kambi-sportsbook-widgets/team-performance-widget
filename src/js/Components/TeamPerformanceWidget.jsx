import React from 'react';
import { widgetModule, translationModule } from 'widget-core-library';
import DetailedEvent from './DetailedEvent';
import DetailedEvents from './DetailedEvents';
import Event from './Event';
import Events from './Events';
import Team from './Team';

/**
 *
 * @type {number}
 */
const HEADER_HEIGHT = 56;

/**
 * Collapsed item plus half bottom border
 * @type {number}
 */
const COMPACT_VIEW_TEAM_INFO_HEIGHT = 74 + 6;

/**
 * Table line height + margin bottom
 * @type {number}
 */
const TABLE_LINE_HEIGHT = 24 + 8;

/**
 * A widget to display the performance indicators of teams last matches.
 */
class TeamPerformanceWidget extends React.Component {

   /**
    * Constructs.
    * @param {object} props Widget properties
    */
   constructor(props) {
      super(props);

      widgetModule.enableWidgetTransition(true);

      this.state = {
         detailed: props.teams.map(() => false)
      };
   }

   /**
    * Called before mounting the widget.
    */
   componentWillMount() {
      document.getElementsByTagName('body')[0].style.display = 'block';
      this.adjustHeight();
   }

   /**
    * Called on attributes change.
    * @param {object} nextProps New attributes
    */
   componentWillReceiveProps(nextProps) {
      this.setState({ detailed: nextProps.teams.map(() => false) });
   }

   /**
    * Called after updating component's DOM.
    */
   componentDidUpdate() {
      this.adjustHeight();
   }

   /**
    * Calculates and updates widget height based on actual content.
    */
   adjustHeight() {
      const contentHeight = this.props.teams.reduce((contentHeight, team, i) => {
         return contentHeight
            + COMPACT_VIEW_TEAM_INFO_HEIGHT
            + (this.state.detailed[i] ? team.lastEvents.length * TABLE_LINE_HEIGHT + 20 : 0);
      }, HEADER_HEIGHT + 1);

      widgetModule.setWidgetHeight(contentHeight);
   }

   /**
    * Handles click on details button.
    * @param {number} teamIdx Team index
    * @param {object} event onChange event
    */
   detailsClickHandler(teamIdx, event) {
      this.state.detailed[teamIdx] = event.target.checked;
      this.adjustHeight();
   }

   /**
    * Creates widget template.
    * @returns {XML}
    */
   render() {
      const t = translationModule.getTranslation.bind(translationModule);

      return (
         <div className="KambiWidget-card-background-color KambiWidget-card-text-color l-flexbox l-vertical kw-wrapper KambiWidget-card-border-color">
            <header className="KambiWidget-font kw-header l-flexbox l-align-center l-pt-16 l-pb-16 l-pl-16">
               {t(this.props.title)}
            </header>
            <main className="KambiWidget-font l-flexbox l-vertical l-flexed l-pack-start l-mb-12">
               {this.props.teams.map((team, i) => {
                  return (
                     <Team key={team.id} team={team} clickHandler={this.detailsClickHandler.bind(this, i)}>
                        <Events>
                           {team.lastEvents.map(event => <Event key={event.start} event={event} />)}
                        </Events>
                        <DetailedEvents>
                           {team.lastEvents.map(event => <DetailedEvent key={event.start} event={event} />)}
                        </DetailedEvents>
                     </Team>
                  );
               })}
            </main>
         </div>
      );
   }
}

TeamPerformanceWidget.propTypes = {
   /**
    * Array of team entities
    */
   teams: React.PropTypes.arrayOf(React.PropTypes.object).isRequired,
   /**
    * Widget's title
    */
   title: React.PropTypes.string.isRequired
};

export default TeamPerformanceWidget;
