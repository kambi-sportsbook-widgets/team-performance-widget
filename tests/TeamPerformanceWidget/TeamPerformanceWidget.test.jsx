/* eslint-env jest */
import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import ReactTestRenderer from 'react-test-renderer';
import { widgetModule } from 'kambi-widget-core-library';
import { mount, shallow } from 'enzyme';
import TeamPerformanceWidget from '../../src/js/Components/TeamPerformanceWidget';
import enzymeToJson from 'enzyme-to-json';
import mockData from '../mockData.json';

let renderer;

jest.mock('kambi-widget-core-library', () => ({
   widgetModule: {
      adaptWidgetHeight: jest.fn(),
   },
   translationModule: {
      getTranslation: jest.fn()
   }
}));


describe('TeamPerformanceWidget DOM rendering', () => {
   beforeEach(() => {
      // enabling fake setTimeout/setInterval etc
      jest.useFakeTimers();

      renderer = new ReactShallowRenderer();
      window.innerWidth = 1024;
   });

   afterEach(() => {
      jest.useRealTimers();
   });

   it('renders correctly', () => {
      expect(renderer.render(
         <TeamPerformanceWidget
            title='theTitle'
            participants={mockData}
         />
      )).toMatchSnapshot();
   });

   it('resize testing', () => {
      // using ReactTestRenderer to test interactions
      // ReactTestRenderer does deep rendering
      // ReactTestUtils does not actually mount the object, so it does not work for this test
      const tree = ReactTestRenderer.create(
         <TeamPerformanceWidget
            title='theTitle'
            participants={mockData}
         />
      );

      widgetModule.adaptWidgetHeight.mockClear();

      const tmpInnerWidth = window.innerWidth;
      window.innerWidth = 480;
      window.dispatchEvent(new Event('resize'));

      expect(window.setTimeout).toHaveBeenCalledTimes(1)
      jest.runAllTimers();
      expect(widgetModule.adaptWidgetHeight).toHaveBeenCalledTimes(1);
   });
});


describe('TeamPerformanceWidget interaction', () => {

   beforeEach(() => {
      renderer = new ReactShallowRenderer();
      window.innerWidth = 1024;
   });

   it('handles switching to detailed view and back', () => {
      const tpi = shallow(
         <TeamPerformanceWidget
            title='theTitle'
            participants={mockData}
         />
      );

      // Detailed view off
      expect(enzymeToJson(tpi)).toMatchSnapshot();

      tpi.find('.KambiWidget-card-background-color').simulate('click', {});

      // Detailed view on
      expect(enzymeToJson(tpi)).toMatchSnapshot();

      tpi.find('.KambiWidget-card-background-color').simulate('click', {});

      // Detailed view off
      expect(enzymeToJson(tpi)).toMatchSnapshot();
   });

});
