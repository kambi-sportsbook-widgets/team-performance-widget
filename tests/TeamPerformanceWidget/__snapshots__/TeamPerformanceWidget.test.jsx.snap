// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TeamPerformanceWidget DOM rendering renders correctly 1`] = `
<div
  className="KambiWidget-card-background-color"
  onClick={[Function]}
>
  <Main
    title="theTitle"
  >
    <Summary
      border={false}
      participants={
        Array [
          Object {
            "id": 1000053608,
            "lastEvents": Array [
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 0,
                "homeName": "1 FSV Mainz 05",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-02-10T19:30:00Z",
              },
              Object {
                "awayName": "Werder Bremen",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-05T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 2,
                "homeName": "VfL Wolfsburg",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-28T14:30:00Z",
              },
              Object {
                "awayName": "1899 Hoffenheim",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 0,
                "result": "lose",
                "start": "2017-01-21T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 1,
                "homeName": "Hannover 96",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-12T15:00:00Z",
              },
              Object {
                "awayName": "AZ Alkmaar",
                "awayScore": 1,
                "homeName": "FC Augsburg",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-07T14:00:00Z",
              },
            ],
            "name": "FC Augsburg",
          },
          Object {
            "id": 1000000113,
            "lastEvents": Array [
              Object {
                "awayName": "Eintracht Frankfurt",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-11T14:30:00Z",
              },
              Object {
                "awayName": "Bayer Leverkusen",
                "awayScore": 0,
                "homeName": "Hamburger SV",
                "homeScore": 1,
                "result": "lose",
                "start": "2017-02-03T19:30:00Z",
              },
              Object {
                "awayName": "Monchengladbach",
                "awayScore": 3,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-01-28T17:30:00Z",
              },
              Object {
                "awayName": "Hertha BSC",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-01-22T14:30:00Z",
              },
              Object {
                "awayName": "VfL Bochum",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "win",
                "start": "2017-01-17T13:00:00Z",
              },
              Object {
                "awayName": "Atletico Mineiro MG",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-12T00:00:00Z",
              },
            ],
            "name": "Bayer Leverkusen",
          },
        ]
      }
    />
  </Main>
</div>
`;

exports[`TeamPerformanceWidget interaction handles switching to detailed view and back 1`] = `
<div
  className="KambiWidget-card-background-color"
  onClick={[Function]}
>
  <Main
    title="theTitle"
  >
    <Summary
      border={false}
      participants={
        Array [
          Object {
            "id": 1000053608,
            "lastEvents": Array [
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 0,
                "homeName": "1 FSV Mainz 05",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-02-10T19:30:00Z",
              },
              Object {
                "awayName": "Werder Bremen",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-05T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 2,
                "homeName": "VfL Wolfsburg",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-28T14:30:00Z",
              },
              Object {
                "awayName": "1899 Hoffenheim",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 0,
                "result": "lose",
                "start": "2017-01-21T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 1,
                "homeName": "Hannover 96",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-12T15:00:00Z",
              },
              Object {
                "awayName": "AZ Alkmaar",
                "awayScore": 1,
                "homeName": "FC Augsburg",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-07T14:00:00Z",
              },
            ],
            "name": "FC Augsburg",
          },
          Object {
            "id": 1000000113,
            "lastEvents": Array [
              Object {
                "awayName": "Eintracht Frankfurt",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-11T14:30:00Z",
              },
              Object {
                "awayName": "Bayer Leverkusen",
                "awayScore": 0,
                "homeName": "Hamburger SV",
                "homeScore": 1,
                "result": "lose",
                "start": "2017-02-03T19:30:00Z",
              },
              Object {
                "awayName": "Monchengladbach",
                "awayScore": 3,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-01-28T17:30:00Z",
              },
              Object {
                "awayName": "Hertha BSC",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-01-22T14:30:00Z",
              },
              Object {
                "awayName": "VfL Bochum",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "win",
                "start": "2017-01-17T13:00:00Z",
              },
              Object {
                "awayName": "Atletico Mineiro MG",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-12T00:00:00Z",
              },
            ],
            "name": "Bayer Leverkusen",
          },
        ]
      }
    />
  </Main>
</div>
`;

exports[`TeamPerformanceWidget interaction handles switching to detailed view and back 2`] = `
<div
  className="KambiWidget-card-background-color"
  onClick={[Function]}
>
  <Main
    title="theTitle"
  >
    <Summary
      border={true}
      participants={
        Array [
          Object {
            "id": 1000053608,
            "lastEvents": Array [
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 0,
                "homeName": "1 FSV Mainz 05",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-02-10T19:30:00Z",
              },
              Object {
                "awayName": "Werder Bremen",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-05T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 2,
                "homeName": "VfL Wolfsburg",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-28T14:30:00Z",
              },
              Object {
                "awayName": "1899 Hoffenheim",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 0,
                "result": "lose",
                "start": "2017-01-21T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 1,
                "homeName": "Hannover 96",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-12T15:00:00Z",
              },
              Object {
                "awayName": "AZ Alkmaar",
                "awayScore": 1,
                "homeName": "FC Augsburg",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-07T14:00:00Z",
              },
            ],
            "name": "FC Augsburg",
          },
          Object {
            "id": 1000000113,
            "lastEvents": Array [
              Object {
                "awayName": "Eintracht Frankfurt",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-11T14:30:00Z",
              },
              Object {
                "awayName": "Bayer Leverkusen",
                "awayScore": 0,
                "homeName": "Hamburger SV",
                "homeScore": 1,
                "result": "lose",
                "start": "2017-02-03T19:30:00Z",
              },
              Object {
                "awayName": "Monchengladbach",
                "awayScore": 3,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-01-28T17:30:00Z",
              },
              Object {
                "awayName": "Hertha BSC",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-01-22T14:30:00Z",
              },
              Object {
                "awayName": "VfL Bochum",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "win",
                "start": "2017-01-17T13:00:00Z",
              },
              Object {
                "awayName": "Atletico Mineiro MG",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-12T00:00:00Z",
              },
            ],
            "name": "Bayer Leverkusen",
          },
        ]
      }
    />
    <Detailed
      participants={
        Array [
          Object {
            "id": 1000053608,
            "lastEvents": Array [
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 0,
                "homeName": "1 FSV Mainz 05",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-02-10T19:30:00Z",
              },
              Object {
                "awayName": "Werder Bremen",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-05T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 2,
                "homeName": "VfL Wolfsburg",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-28T14:30:00Z",
              },
              Object {
                "awayName": "1899 Hoffenheim",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 0,
                "result": "lose",
                "start": "2017-01-21T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 1,
                "homeName": "Hannover 96",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-12T15:00:00Z",
              },
              Object {
                "awayName": "AZ Alkmaar",
                "awayScore": 1,
                "homeName": "FC Augsburg",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-07T14:00:00Z",
              },
            ],
            "name": "FC Augsburg",
          },
          Object {
            "id": 1000000113,
            "lastEvents": Array [
              Object {
                "awayName": "Eintracht Frankfurt",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-11T14:30:00Z",
              },
              Object {
                "awayName": "Bayer Leverkusen",
                "awayScore": 0,
                "homeName": "Hamburger SV",
                "homeScore": 1,
                "result": "lose",
                "start": "2017-02-03T19:30:00Z",
              },
              Object {
                "awayName": "Monchengladbach",
                "awayScore": 3,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-01-28T17:30:00Z",
              },
              Object {
                "awayName": "Hertha BSC",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-01-22T14:30:00Z",
              },
              Object {
                "awayName": "VfL Bochum",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "win",
                "start": "2017-01-17T13:00:00Z",
              },
              Object {
                "awayName": "Atletico Mineiro MG",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-12T00:00:00Z",
              },
            ],
            "name": "Bayer Leverkusen",
          },
        ]
      }
    />
  </Main>
</div>
`;

exports[`TeamPerformanceWidget interaction handles switching to detailed view and back 3`] = `
<div
  className="KambiWidget-card-background-color"
  onClick={[Function]}
>
  <Main
    title="theTitle"
  >
    <Summary
      border={false}
      participants={
        Array [
          Object {
            "id": 1000053608,
            "lastEvents": Array [
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 0,
                "homeName": "1 FSV Mainz 05",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-02-10T19:30:00Z",
              },
              Object {
                "awayName": "Werder Bremen",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-05T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 2,
                "homeName": "VfL Wolfsburg",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-28T14:30:00Z",
              },
              Object {
                "awayName": "1899 Hoffenheim",
                "awayScore": 2,
                "homeName": "FC Augsburg",
                "homeScore": 0,
                "result": "lose",
                "start": "2017-01-21T14:30:00Z",
              },
              Object {
                "awayName": "FC Augsburg",
                "awayScore": 1,
                "homeName": "Hannover 96",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-12T15:00:00Z",
              },
              Object {
                "awayName": "AZ Alkmaar",
                "awayScore": 1,
                "homeName": "FC Augsburg",
                "homeScore": 1,
                "result": "draw",
                "start": "2017-01-07T14:00:00Z",
              },
            ],
            "name": "FC Augsburg",
          },
          Object {
            "id": 1000000113,
            "lastEvents": Array [
              Object {
                "awayName": "Eintracht Frankfurt",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-02-11T14:30:00Z",
              },
              Object {
                "awayName": "Bayer Leverkusen",
                "awayScore": 0,
                "homeName": "Hamburger SV",
                "homeScore": 1,
                "result": "lose",
                "start": "2017-02-03T19:30:00Z",
              },
              Object {
                "awayName": "Monchengladbach",
                "awayScore": 3,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "lose",
                "start": "2017-01-28T17:30:00Z",
              },
              Object {
                "awayName": "Hertha BSC",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 3,
                "result": "win",
                "start": "2017-01-22T14:30:00Z",
              },
              Object {
                "awayName": "VfL Bochum",
                "awayScore": 1,
                "homeName": "Bayer Leverkusen",
                "homeScore": 2,
                "result": "win",
                "start": "2017-01-17T13:00:00Z",
              },
              Object {
                "awayName": "Atletico Mineiro MG",
                "awayScore": 0,
                "homeName": "Bayer Leverkusen",
                "homeScore": 1,
                "result": "win",
                "start": "2017-01-12T00:00:00Z",
              },
            ],
            "name": "Bayer Leverkusen",
          },
        ]
      }
    />
  </Main>
</div>
`;
